rcl/cmake/configure_rcl.cmake

```diff
@@ -0,0 +1,73 @@
+# Copyright 2019 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Configures ros client library with custom settings.
+# The custom settings are all related to library symbol visibility, see:
+# https://gcc.gnu.org/wiki/Visibility
+# http://www.ibm.com/developerworks/aix/library/au-aix-symbol-visibility/
+#
+# Below code is heavily referenced from a similar functionality in rmw:
+# https://github.com/ros2/rmw/blob/master/rmw/cmake/configure_rmw_library.cmake
+#
+# :param library_target: the library target
+# :type library_target: string
+# :param LANGUAGE: Optional flag for the language of the library.
+#   Allowed values are "C" and "CXX". The default is "CXX".
+# :type LANGUAGE: string
+#
+# @public
+#
+macro(configure_rcl library_target)
+  cmake_parse_arguments(_ARG "" "LANGUAGE" "" ${ARGN})
+  if(_ARG_UNPARSED_ARGUMENTS)
+    message(FATAL_ERROR "configure_rcl() called with unused arguments: ${_ARG_UNPARSED_ARGUMENTS}")
+  endif()
+
+  if(NOT _ARG_LANGUAGE)
+    set(_ARG_LANGUAGE "CXX")
+  endif()
+
+  if(_ARG_LANGUAGE STREQUAL "C")
+    # Set the visibility to hidden by default if possible
+    if(CMAKE_C_COMPILER_ID STREQUAL "GNU" OR CMAKE_C_COMPILER_ID MATCHES "Clang")
+      # Set the visibility of symbols to hidden by default for gcc and clang
+      # (this is already the default on Windows)
+      set_target_properties(${library_target}
+        PROPERTIES
+        COMPILE_FLAGS "-fvisibility=hidden"
+      )
+    endif()
+
+  elseif(_ARG_LANGUAGE STREQUAL "CXX")
+    # Set the visibility to hidden by default if possible
+    if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU" OR CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+      # Set the visibility of symbols to hidden by default for gcc and clang
+      # (this is already the default on Windows)
+      set_target_properties(${library_target}
+        PROPERTIES
+        COMPILE_FLAGS "-fvisibility=hidden -fvisibility-inlines-hidden"
+      )
+    endif()
+
+  else()
+    message(FATAL_ERROR "configure_rcl() called with unsupported LANGUAGE: '${_ARG_LANGUAGE}'")
+  endif()
+
+  if(WIN32)
+    # Causes the visibility macros to use dllexport rather than dllimport
+    # which is appropriate when building the dll but not consuming it.
+    target_compile_definitions(${library_target}
+      PRIVATE "RCL_BUILDING_DLL")
```

Comment:
The reason it works for the rmw function you're borrowing this code from is that the headers are in the one package, and there are several packages which provide the C++ code (actually create the library target) that fulfills  the rmw interface.

I'd recommend refactoring the rmw function to extract just the `-fvisibility=hidden` stuff, maybe to a macro called `rmw_set_symbol_visibility_hidden(library_target [LANGUAGE {C, CXX}])`, then call that new macro from within `configure_rmw_library` and this macro.

Reply:
I have actually removed the above target definition code in the most recent update on this PR. This is mainly because this definition is already set as part of CMakeList: https://github.com/ros2/rcl/blob/master/rcl/CMakeLists.txt#L73-L75

So, the function is just setting the visibility now. Let me know if we should change the macro name to "rcl_set_symbol_visibility_hidden" or let it stay generic "configure_rcl" so that future use cases can be added here

Reply:
I'd say that I prefer something narrow in scope like `rcl_set_symbol_visibility_hidden`, but if you want to call that from inside of a general function, that would be fine.

For the currently proposed generic function, why not `configure_rcl_library` so it matches the existing `configure_rmw_library`?

Honestly, the only reason `configure_rmw_library` exists is that there are multiple implementations of the same set of headers `rmw` headers, which in my opinion is a special use case. It's not clear to me that it is appropriate to have a generic function for `rcl*` libraries/packages, as it is for `rmw`. It may or may not be appropriate to hide symbols and stuff like that on different packages. For example, I don't if it should be called on `rclpy`'s library or not. For that reason, I'd prefer just "small sharp tool" like functions that are called as necessary in each package that needs them.

If the number of packages that need to called these functions is larger than I'm thinking then we can reconsider having a generic function, but in that case I think the name should be different as `configure_rcl` isn't very descriptive.

Reply:
Makes sense, I will change the name of the macro. I was avoiding configure_rcl_libray as rcl itself is ros client library :P

Reply:
Updated with the name changes

Reply:
> Makes sense, I will change the name of the macro. I was avoiding configure_rcl_libray as rcl itself is ros client library :P

That's fair, though I read more like "configure a library target which is also an 'rcl'". Maybe a more "correct"/less ambiguous (but not necessarily better) name would be `config_library_for_rcl` or `rcl_configure_library`.